Add return statement to the JIT AST

apaszke opened the pull request:
This also makes the syntax recognized by the parser compatible with Python syntax, so it will allow us to reuse Python functions for script tests as well.

apaszke committed:
Add return statement to the JIT AST

zdevito commented on torch/csrc/jit/script/parser.h:
```diff
@@ -297,12 +285,17 @@ struct Parser {
-        return c(TK_GLOBAL, range, std::move(idents));
+        auto idents = parseList(TK_NOTHING, ',', TK_NOTHING, &Parser::parseIdent);
+        if (idents.empty())
+          throw ErrorReport(range) << "global statement needs a list of variable names";
```
How can this happen? Assuming there are no TK_NOTHING tokens in the stream, then there is at least one element in this list or an exception is thrown.

apaszke replied:
Right. Now that I read the body again, it uses a `do ... while` instead of `while`, so it has to be non-empty when end is `TK_NOTHING`.

zdevito reviewed (approved):
Looks good! one minor question

apaszke committed:
Review comments + TK_OPTION -> TK_NOTHING

apaszke commented:
Fixed the comment and changed all uses of `TK_OPTION` to `TK_NOTHING`. They have been used to denote similar things, but saying that a list ends with `TK_OPTION` is weird.

zdevito commented on torch/csrc/jit/script/parser.h:
```diff
@@ -228,17 +228,17 @@ struct Parser {
     // Now we *may* have an expression.
     if (L.cur().kind != ']') {
-      second = c(TK_OPTION, range, {parseExp()});
+      second = c(TK_NOTHING, range, {parseExp()});
```
Huh? This was correct before. This is an option node with a single value _present_. It is certainly not nothing. TK_OPTION has a single subtree in the event a value is present, and no subtrees otherwise. Can you change this back?

apaszke replied:
Good point. I changed it without reading into it, based on the comment in tree views. This means that the Maybe node was broken since the beginning. Will fix.
